### PR TITLE
[WIP] ENH Pytorch 3D Back/Frontend 

### PR DESCRIPTION
**Pull request by edouardoyallon**
adding torch backend

### CODE REVIEW

File: kymatio/scattering3d/backend/torch_backend.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+    """
+    if not A.is_contiguous():
+        warnings.warn("cdgmm3d: tensor A is converted to a contiguous array.")
+        A = A.contiguous()
+    if not B.is_contiguous():
+        warnings.warn("cdgmm3d: tensor B is converted to a contiguous array.")
+        B = B.contiguous()
+
+    if A.size()[-4:] != B.size():
+        raise RuntimeError('The tensors are not compatible for multiplication!')
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/backend/torch_backend.py]: Remove exclamation mark.

File: kymatio/scattering3d/backend/torch_backend.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+    if not iscomplex(A) or not iscomplex(B):
+        raise TypeError('The input, filter and output should be complex')
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/backend/torch_backend.py]: Punctuation.

File: kymatio/scattering3d/backend/torch_backend.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+        raise RuntimeError('A and B should be same type!')
+
+    if A.device.type != B.device.type:
+        raise TypeError('A and B must be both on GPU or both on CPU.')
+
+    if A.device.type == 'cuda':
+        if A.device.index != B.device.index:
+            raise TypeError('A and B must be on the same GPU!')
+
+    C = A.new(A.size())
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/backend/torch_backend.py]: Shape.

File: kymatio/scattering3d/backend/torch_backend.py [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+    C[..., 0] = A[..., 0] * B[..., 0] - A[..., 1] * B[..., 1]
+    C[..., 1] = A[..., 0] * B[..., 1] + A[..., 1] * B[..., 0]
+
+    return C if not inplace else A.copy_(C)
+
+
+def finalize(s_order_1, s_order_2, max_order):
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/backend/torch_backend.py]: Concatenate?

File: kymatio/scattering3d/frontend/base_frontend.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
+from ...frontend.base_frontend import ScatteringBase
+
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/frontend/base_frontend.py]: Remove space.

File: kymatio/scattering3d/frontend/base_frontend.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+from ..filter_bank import solid_harmonic_filter_bank, gaussian_filter_bank
+
+
+
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/frontend/base_frontend.py]: Remove space.

File: kymatio/scattering3d/frontend/__init__.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        if 'frontend' not in kwargs:
+            warnings.warn("Torch frontend is currently the default, but NumPy will become the default in the next"
+                          " version.", PendingDeprecationWarning)
+            frontend = 'torch'
+        else:
+            frontend = kwargs['frontend'].lower()
+            kwargs.pop('frontend')
+
+        try:
+            module = importlib.__import__(frontend + '_frontend', globals(), locals(), [], 1)
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/frontend/__init__.py]: Import Module.

File: kymatio/scattering3d/frontend/base_frontend.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+             Spatial support of the padded input.
+
+        Notes
+        -----
+        The design of the filters is optimized for the value L = 8.
+
+        pre_pad is particularly useful when cropping bigger images because
+        this does not introduce border effects inherent to padding.
+        """
+
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/frontend/base_frontend.py]: Remove space.

File: kymatio/scattering3d/frontend/base_frontend.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    def build(self):
+        self.M, self.N, self.O = self.shape
+
+        if not self.backend:
+            from ..backend.torch_backend import backend  # is imported like a module and not a class?
+            self.backend = backend
+        elif self.backend.name[0:5] != 'torch':
+            raise RuntimeError('This backend is not supported.')
+
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/frontend/base_frontend.py]: Remove space.

File: kymatio/scattering3d/frontend/base_frontend.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+
+    def create_filters(self):
+        self.filters = solid_harmonic_filter_bank(
+            self.M, self.N, self.O, self.J, self.L, self.sigma_0)
+
+        self.gaussian_filters = gaussian_filter_bank(
+            self.M, self.N, self.O, self.J + 1, self.sigma_0)
+
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/frontend/base_frontend.py]: Add space.

File: kymatio/scattering3d/frontend/torch_frontend.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
+# Authors: Louis Thiry, Georgios Exarchakis
+# Scientific Ancestry: Louis Thiry, Georgios Exarchakis, Matthew Hirn, Michael Eickenberg
+
+from ...frontend.torch_frontend import ScatteringTorch
+import torch
+from .base_frontend import ScatteringBase3D
+
+from kymatio.scattering3d.core.scattering3d import scattering3d
+
+
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/frontend/torch_frontend.py]: Remove line.

File: kymatio/scattering3d/frontend/torch_frontend.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
+# Authors: Louis Thiry, Georgios Exarchakis
+# Scientific Ancestry: Louis Thiry, Georgios Exarchakis, Matthew Hirn, Michael Eickenberg
+
+from ...frontend.torch_frontend import ScatteringTorch
+import torch
+from .base_frontend import ScatteringBase3D
+
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/frontend/torch_frontend.py]: Remove line

File: kymatio/scattering3d/frontend/torch_frontend.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        g = torch.zeros(self.gaussian_filters.shape + (2,))
+        g[..., 0] = torch.from_numpy(self.gaussian_filters.real)
+        self.gaussian_filters = g
+        self.register_buffer('tensor_gaussian_filter', self.gaussian_filters)
+
+
+
+
+
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/frontend/torch_frontend.py]: Remove.

File: kymatio/scattering3d/frontend/torch_frontend.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    def scattering(self, input_array):
+        buffer_dict = dict(self.named_buffers())
+        for k in range(len(self.filters)):
+            self.filters[k] = buffer_dict['tensor' + str(k)]
+
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/frontend/torch_frontend.py]: Remove line.

File: kymatio/scattering3d/tests/test_torch_scattering3d.py [hunk 15]
```diff
@@ -0,0 +1,294 @@
+""" This script will test the submodules used by the scattering module"""
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/tests/test_torch_scattering3d.py]: Fix the spaces/empty lines.

File: kymatio/scattering3d/backend/torch_backend.py [hunk 16]
```diff
@@ -0,0 +1,326 @@
+import torch
+import warnings
+
+BACKEND_NAME = 'torch'
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/backend/torch_backend.py]: Is this variable used??

**Comment by edouardoyallon** [comment]
@janden if you have time tonight!

File: kymatio/scattering3d/backend/torch_backend.py [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+            Tensor the same size as input_array. modulus[..., 0] holds the
+            result of the complex modulus, modulus[..., 1] = 0.
+
+    """
+    modulus = torch.zeros_like(input_array)
+    modulus[..., 0] += torch.sqrt((input_array ** 2).sum(-1))
+    return modulus
+
+
+def modulus_rotation(x, module):
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/scattering3d/backend/torch_backend.py]: Lets make an issue.

File: kymatio/scattering3d/backend/torch_backend.py [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+            The result of input_array \\star phi_J downsampled by a factor J.
+
+    """
+    low_pass = filter[J]
+    convolved_input = cdgmm3d(input_array, low_pass)
+    convolved_input = fft(convolved_input, inverse=True)
+    return subsample(convolved_input, J)
+
+
+def _compute_local_scattering_coefs(input_array, filter, j, points):
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/scattering3d/backend/torch_backend.py]: @eickenberg Think you'd be able to make some tests for this?

File: kymatio/scattering3d/backend/torch_backend.py [hunk 19]
```diff
@@ -0,0 +1,326 @@
...
+    convolved_input = fft(convolved_input, inverse=True)
+    for i in range(input_array.size(0)):
+        for j in range(points[i].size(0)):
+            x, y, z = points[i, j, 0], points[i, j, 1], points[i, j, 2]
+            local_coefs[i, j, 0] = convolved_input[
+                i, int(x), int(y), int(z), 0]
+    return local_coefs
+
+
+def subsample(input_array, j):
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/scattering3d/backend/torch_backend.py]: @eickenberg Think you'd be able to make some tests for this?

File: kymatio/scattering3d/backend/torch_backend.py [hunk 20]
```diff
@@ -0,0 +1,325 @@
...
+        Returns
+        -------
+        out : tensor
+            Downsampled tensor.
+
+    """
+    return input_array[..., ::2 ** j, ::2 ** j, ::2 ** j, :].contiguous()
+
+
+def compute_integrals(input_array, integral_powers):
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/backend/torch_backend.py]: @eickenberg Same.

File: kymatio/scattering3d/utils.py [hunk 21]
```diff
@@ -148,11 +101,10 @@ def _apply_filters(filters, fn):
...
         fn: a function to apply on the parameters
-
         Returns
         -------
         filters: the filters modified
     """
     for k in range(len(filters)):
         filters[k] = fn(filters[k])
-    return filters
+    return filters
```
Comment (janden) [inline kymatio/scattering3d/utils.py]: Add newline.

File: kymatio/scattering3d/frontend/__init__.py [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+import warnings
+import importlib
+
+
+class HarmonicScattering3D(object):
+    def __init__(self, *args, **kwargs):
+        frontend_suffixes = {'torch' : 'Torch', 'numpy' : 'NumPy', 'tensorflow'
+                : 'TensorFlow'}
+        if 'frontend' not in kwargs:
+            warnings.warn("Torch frontend is currently the default, but NumPy will become the default in the next"
```
Comment (eickenberg) [inline kymatio/scattering3d/frontend/__init__.py]: make line shorter

File: kymatio/scattering3d/frontend/__init__.py [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+
+            class_name = self.__class__.__name__
+            class_name = (class_name[:-2] + frontend + class_name[-2:])
+
+            self.__class__ = getattr(module, class_name)
+            self.__init__(*args, **kwargs)
+        except Exception as e:
+            raise e from RuntimeError('\nThe frontend \'' + frontend + '\' could not be correctly imported.')
+
+        logging.info('The hardmonic 3D frontend ' + frontend + ' was imported.')
```
Comment (eickenberg) [inline kymatio/scattering3d/frontend/__init__.py]: typo

File: kymatio/scattering3d/frontend/base_frontend.py [hunk 24]
```diff
@@ -0,0 +1,110 @@
+from ...frontend.base_frontend import ScatteringBase
+from ..filter_bank import solid_harmonic_filter_bank, gaussian_filter_bank
+
+
+class ScatteringBase3D(ScatteringBase):
+    """ Abstract module implementing the scattering transform in 2D. Actual
```
Comment (eickenberg) [inline kymatio/scattering3d/frontend/base_frontend.py]: 3D

File: kymatio/scattering3d/frontend/base_frontend.py [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+        implementations of the Scattering2D inherits from its methods and
```
Comment (eickenberg) [inline kymatio/scattering3d/frontend/base_frontend.py]: 3D

File: kymatio/scattering3d/frontend/base_frontend.py [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+
+        where * denotes the convolution (in space), phi_J is a lowpass
+        filter, psi^1_lambda is a family of bandpass
+        filters and psi^2_mu is another family of bandpass filters.
+        Only Morlet filters are used in this implementation.
+        Convolutions are efficiently performed in the Fourier domain.
+
+        Example
+        -------
+            # 1) Define a Scattering2D object as:
```
Comment (eickenberg) [inline kymatio/scattering3d/frontend/base_frontend.py]: 3D

File: kymatio/scattering3d/frontend/base_frontend.py [hunk 27]
```diff
@@ -0,0 +1,110 @@
...
+                s = Scattering2D(J, shape=(M, N))
```
Comment (eickenberg) [inline kymatio/scattering3d/frontend/base_frontend.py]: 3D

File: kymatio/scattering3d/frontend/torch_frontend.py [hunk 28]
```diff
@@ -0,0 +1,154 @@
...
+        integration. Used with method == 'standard', method == 'integral'
+    """
+    def __init__(self, J, shape, L=3, sigma_0=1, max_order=2, rotation_covariant=True, method='standard', points=None,
+                 integral_powers=(0.5, 1., 2.), backend='torch'):
+        ScatteringTorch.__init__(self)
+        ScatteringBase3D.__init__(self, J, shape, L, sigma_0, max_order, rotation_covariant, method, points,
+                 integral_powers, backend)
+        ScatteringBase3D._instantiate_backend(self, 'kymatio.scattering3d.backend.')
+        ScatteringBase3D.build(self)
+        ScatteringBase3D.create_filters(self)
```
Comment (eickenberg) [inline kymatio/scattering3d/frontend/torch_frontend.py]: this goes into build

File: kymatio/scattering3d/frontend/torch_frontend.py [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+
+        self.register_filters()
```
Comment (eickenberg) [inline kymatio/scattering3d/frontend/torch_frontend.py]: this goes into build

File: kymatio/scattering3d/frontend/torch_frontend.py [hunk 30]
```diff
@@ -0,0 +1,154 @@
...
+    integral_powers: array-like
+        List of exponents to the power of which moduli are raised before
+        integration. Used with method == 'standard', method == 'integral'
+    """
+    def __init__(self, J, shape, L=3, sigma_0=1, max_order=2, rotation_covariant=True, method='standard', points=None,
+                 integral_powers=(0.5, 1., 2.), backend='torch'):
+        ScatteringTorch.__init__(self)
+        ScatteringBase3D.__init__(self, J, shape, L, sigma_0, max_order, rotation_covariant, method, points,
+                 integral_powers, backend)
+        ScatteringBase3D._instantiate_backend(self, 'kymatio.scattering3d.backend.')
```
Comment (eickenberg) [inline kymatio/scattering3d/frontend/torch_frontend.py]: this goes into build

File: kymatio/scattering3d/frontend/torch_frontend.py [hunk 31]
```diff
@@ -0,0 +1,154 @@
...
+        g[..., 0] = torch.from_numpy(self.gaussian_filters.real)
+        self.gaussian_filters = g
+        self.register_buffer('tensor_gaussian_filter', self.gaussian_filters)
+
+    def scattering(self, input_array):
+        buffer_dict = dict(self.named_buffers())
+        for k in range(len(self.filters)):
+            self.filters[k] = buffer_dict['tensor' + str(k)]
+
+        methods = ['standard', 'local', 'integral']
```
Comment (eickenberg) [inline kymatio/scattering3d/frontend/torch_frontend.py]: fyi local wasn't in any publication

File: kymatio/scattering3d/frontend/torch_frontend.py [hunk 32]
```diff
@@ -0,0 +1,154 @@
...
+        if (not self.method in methods):
+            raise (ValueError('method must be in {}'.format(methods)))
+        if self.method == 'integral': \
+                self.averaging = lambda x, j: self.backend.compute_integrals(self.backend.fft(x, inverse=True)[..., 0],
+                                                                             self.integral_powers)
+        elif self.method == 'local':
+            self.averaging = lambda x, j: \
+                self.backend._compute_local_scattering_coefs(x,
+                                                             self.tensor_gaussian_filter, j, self.points)
+        elif self.method == 'standard':
```
Comment (eickenberg) [inline kymatio/scattering3d/frontend/torch_frontend.py]: this option might mean that we need to handle our outputs in a way that can keep space dimensions. relevant to the final reshapes

File: kymatio/scattering3d/tests/test_torch_scattering3d.py [hunk 33]
```diff
@@ -0,0 +1,303 @@
...
+from kymatio.scattering3d.utils import generate_weighted_sum_of_gaussians
+
+
+backends = []
+
+try:
+    if torch.cuda.is_available():
+        from skcuda import cublas
+        import cupy
+        from kymatio.scattering3d.backend.torch_skcuda_backend import backend
```
Comment (eickenberg) [inline kymatio/scattering3d/tests/test_torch_scattering3d.py]: will silently fail

File: kymatio/scattering3d/tests/test_torch_scattering3d.py [hunk 34]
```diff
@@ -0,0 +1,303 @@
...
+backends.append(backend)
+
+if torch.cuda.is_available():
+    devices = ['cuda', 'cpu']
+else:
+    devices = ['cpu']
+
+
+def relative_difference(a, b):
+    return np.sum(np.abs(a - b)) / max(np.sum(np.abs(a)), np.sum(np.abs(b)))
```
Comment (eickenberg) [inline kymatio/scattering3d/tests/test_torch_scattering3d.py]: max or min?

File: kymatio/scattering3d/tests/test_torch_scattering3d.py [hunk 35]
```diff
@@ -0,0 +1,303 @@
...
+@pytest.mark.parametrize("backend", backends)
+def test_FFT3d_central_freq_batch(device, backend):
+    # Checked the 0 frequency for the 3D FFT
+    for device in devices:
+        x = torch.zeros(1, 32, 32, 32, 2).float()
+        if device == 'gpu':
+            x = x.cuda()
+        a = x.sum()
+        y = backend.fft(x)
+        c = y[:,0,0,0].sum()
```
Comment (eickenberg) [inline kymatio/scattering3d/tests/test_torch_scattering3d.py]: space after comma

File: kymatio/scattering3d/tests/test_torch_scattering3d.py [hunk 36]
```diff
@@ -0,0 +1,303 @@
...
+        assert (c-a).abs().sum()<1e-6
```
Comment (eickenberg) [inline kymatio/scattering3d/tests/test_torch_scattering3d.py]: spaces around operators

File: kymatio/scattering3d/tests/test_torch_scattering3d.py [hunk 37]
```diff
@@ -0,0 +1,303 @@
...
+def test_cdgmm3d(device, backend, inplace):
+    if backend.name == 'torch' or device != 'cpu':
+        # Not all backends currently implement the inplace variant
+        x = torch.zeros(2, 3, 4, 2).to(device)
+        x[...,0] = 2
+        x[...,1] = 3
+
+        y = torch.zeros_like(x)
+        y[...,0] = 4
+        y[...,1] = 5
```
Comment (eickenberg) [inline kymatio/scattering3d/tests/test_torch_scattering3d.py]: spaces after commas

File: kymatio/scattering3d/tests/test_torch_scattering3d.py [hunk 38]
```diff
@@ -0,0 +1,303 @@
...
+
+        prod = torch.zeros_like(x)
+        prod[...,0] = x[...,0]*y[...,0] - x[...,1]*y[...,1]
+        prod[...,1] = x[...,0]*y[...,1] + x[...,1]*y[...,0]
```
Comment (eickenberg) [inline kymatio/scattering3d/tests/test_torch_scattering3d.py]: spaces after commas

Comment (eickenberg) [reply kymatio/scattering3d/tests/test_torch_scattering3d.py]: also below this

File: kymatio/scattering3d/tests/test_utils_scattering3d.py [hunk 39]
```diff
@@ -0,0 +1,26 @@
...
+    y = sqrt(x**2)
+    assert (y == x).all()
+
+    # Make sure we still don't let in negatives...
+    with pytest.warns(RuntimeWarning) as record:
+        x = np.array([-1, 0, 1])
+        y = sqrt(x)
+    assert "Negative" in record[0].message.args[0]
+
+    # ...unless they are compled numbers!
```
Comment (eickenberg) [inline kymatio/scattering3d/tests/test_utils_scattering3d.py]: complex?

File: kymatio/scattering3d/utils.py [hunk 40]
```diff
@@ -1,97 +1,54 @@
 """Author: Louis Thiry, All rights reserved, 2018."""
```
Comment (eickenberg) [inline kymatio/scattering3d/utils.py]: it's not "All rights reserved".

File: kymatio/scattering3d/utils.py (hunk 21 above)
Comment (eickenberg) [inline kymatio/scattering3d/utils.py]: \n

File: kymatio/scattering3d/backend/torch_backend.py [hunk 41]
```diff
@@ -0,0 +1,326 @@
...
+        Parameters
+        ----------
+        input_array : torch tensor
+            Size (batchsize, M, N, O, 2).
+        filter : torch tensor
+            Size (M, N, O, 2).
+        J : int
+            Low pass scale of phi_J.
+        subsample : function
+            Subsampling function.
```
[the remark this answers is not in the document]
Comment (janden) [reply kymatio/scattering3d/backend/torch_backend.py]: I have no idea.

File: kymatio/scattering3d/frontend/torch_frontend.py (hunk 31 above)
Comment (janden) [reply kymatio/scattering3d/frontend/torch_frontend.py]: Added a comment.

File: kymatio/scattering3d/backend/torch_backend.py (hunk 4 above)
Comment (janden) [reply kymatio/scattering3d/backend/torch_backend.py]: Gone.

File: kymatio/scattering3d/frontend/torch_frontend.py (hunk 31 above)
Comment (janden) [reply kymatio/scattering3d/frontend/torch_frontend.py]: And an issue.

File: kymatio/scattering3d/frontend/torch_frontend.py (hunk 32 above)
Comment (janden) [reply kymatio/scattering3d/frontend/torch_frontend.py]: This is fixed by counting forwards along the axes (not backwards).

File: kymatio/scattering3d/tests/test_torch_scattering3d.py (hunk 33 above)
Comment (janden) [reply kymatio/scattering3d/tests/test_torch_scattering3d.py]: Made an issue.

File: kymatio/scattering3d/backend/torch_backend.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (janden) [reply kymatio/scattering3d/backend/torch_backend.py]: Made an issue. Resolving conversation.

File: kymatio/scattering3d/backend/torch_backend.py (hunk 16 above)
Comment (edouardoyallon) [reply kymatio/scattering3d/backend/torch_backend.py]: Let's keep it for now and leave it later

**Pull request merged**